packages/flutter/lib/src/rendering/proxy_box.dart

```diff
@@ -2355,18 +2371,32 @@ class RenderTransform extends RenderProxyBox {
   void paint(PaintingContext context, Offset offset) {
     if (child != null) {
       final Matrix4 transform = _effectiveTransform!;
-      final Offset? childOffset = MatrixUtils.getAsTranslation(transform);
-      if (childOffset == null) {
-        layer = context.pushTransform(
-          needsCompositing,
-          offset,
-          transform,
-          super.paint,
-          oldLayer: layer as TransformLayer?,
-        );
+      if (filterQuality == null) {
+        final Offset? childOffset = MatrixUtils.getAsTranslation(transform);
+        if (childOffset == null) {
+          layer = context.pushTransform(
+            needsCompositing,
+            offset,
+            transform,
+            super.paint,
+            oldLayer: layer is TransformLayer ? layer as TransformLayer? : null,
+          );
+        } else {
+          super.paint(context, offset + childOffset);
+          layer = null;
+        }
       } else {
-        super.paint(context, offset + childOffset);
-        layer = null;
+        final ui.ImageFilter filter = ui.ImageFilter.matrix(
+          transform.storage,
+          filterQuality: filterQuality!,
+        );
+        if (layer is ImageFilterLayer) {
+          final ImageFilterLayer filterLayer = layer! as ImageFilterLayer;
+          filterLayer.imageFilter = filter;
+        } else {
+          layer = ImageFilterLayer(imageFilter: filter);
+        }
+        context.pushLayer(layer!, super.paint, offset);
```

Comment:
If this layer it going to unconditionally push a layer, then we need to make sure that we set `needsCompositing` to true, or that `alwaysNeedsCompositing` is set to true.

Reply:
Adding/removing a child (via adoptChild) will request an update to the compositing bits. It looks like the update method here will look for a child that requires compositing.

What I think may be missing (and it was missing before my modifications) is that if the transform evaluates to a translation then we avoid a layer and that won't be reflected in the value of `needsCompositing` for this RO or its parents. That issue is not changed here.

Reply:
I'm not sure that's the case. Previously we had:

```dart
if (childOffset == null) {
  layer = context.pushTransform(needsCompositing, ..., oldLayer: layer);
} else {
  super.paint(context, offset + childOffset);
  layer = null;
}
```

This says that if it's not a simple translation (the `if` branch), then `context.pushTransform` will either push a layer or inject a canvas transform, depending on the value of `needsCompositing`. If it's a simple transform (the `else` branch), we don't push layers. So there are 3 possible outcomes:

1. Not simple translation, needs compositing: push layer
2. Not simple translation, needs not compositing: no layer
3. Simple translation: delegate to children

There's no situation when `needsCompositing == false` and we push a layer.

After this PR, we have the following situation: `filterQuality != null`, `needsCompositing == false` => push a layer.

It's OK to push a layer when none of the descendants need compositing, but we still need to set the render object's _own_ `needsCompositing` to `true`. Otherwise, its ancestors will do wrong things.

Reply:
I see. Basically it would only push a transform layer if a child was already planning to composite (push a layer?). So, it was impartial wrt whether it was going to push a layer itself - if no children were planning to, then it didn't. Thus, this layer used to not have any additional requirements as to the compositing bits other than the "standard" pass through from the children.

Now that FQ means we will be pushing a layer, that needs to factor in...

Reply:
I have now updated the alwaysNeedsCompositing handling to match what I think you were saying here.

Reply:
Yep, the latest iteration looks correct.